File: frontend/src/types/activityLog.ts
/**
 * Tipos para el sistema de bitácora inteligente
 */

export interface ActivityLogBase {
  stage_id: number;
  activity_type: 'nueva_actividad' | 'seguimiento' | 'cierre_etapa';
  start_date: string;
  end_date?: string;
  next_follow_up_at?: string;
  status: 'pendiente' | 'en_curso' | 'completada' | 'cancelada';
  actor_type: 'equipo_interno' | 'proveedor' | 'usuario';
  notes?: string;
  dynamic_payload?: Record<string, any>;
}

export interface ActivityLogCreate extends ActivityLogBase {}

export interface ActivityLogUpdate {
  end_date?: string;
  next_follow_up_at?: string;
  status?: 'pendiente' | 'en_curso' | 'completada' | 'cancelada';
  notes?: string;
  dynamic_payload?: Record<string, any>;
}

export interface ActivityLogResponse extends ActivityLogBase {
  id: number;
  development_id: string;
  created_by?: string;
  created_at: string;
  updated_at?: string;
  stage_name?: string;
  stage_code?: string;
}

export interface ActivityLogListResponse {
  activities: ActivityLogResponse[];
  total: number;
  page: number;
  size: number;
}

export interface StageFieldConfig {
  stage_id: number;
  stage_name: string;
  stage_code: string;
  field_config?: {
    fields: any;
    required_fields: string[];
    optional_fields: string[];
  };
  has_dynamic_fields: boolean;
}

// Campos dinámicos por etapa
export interface AprobacionPropuestaFields {
  proposal_id: string;
  proposal_version?: string;
  approver?: string;
  approval_date?: string;
  budget_approved?: number;
}

export interface AnalisisFields {
  side_case_id?: string;
  servicepoint_case_id?: string;
  analyst_assigned?: string;
  scope_analysis?: string;
  requirements_count?: number;
}

export interface InstalacionPruebasFields {
  installer_number: string;
  environment?: string;
  change_window?: string;
  installation_notes?: string;
}

// Configuración de campos por etapa
export const STAGE_FIELD_CONFIGS: Record<string, {
  fields: any;
  required_fields: string[];
  optional_fields: string[];
  display_name: string;
  description: string;
}> = {
  'Aprobación Propuesta': {
    fields: null, // Se define dinámicamente
    required_fields: ['proposal_id'],
    optional_fields: ['proposal_version', 'approver', 'approval_date', 'budget_approved'],
    display_name: 'Aprobación de Propuesta',
    description: 'Gestión de aprobación de propuestas técnicas'
  },
  'Análisis': {
    fields: null,
    required_fields: [],
    optional_fields: ['side_case_id', 'servicepoint_case_id', 'analyst_assigned', 'scope_analysis', 'requirements_count'],
    display_name: 'Análisis',
    description: 'Análisis de requerimientos y casos de negocio'
  },
  'Despliegue (Pruebas)': {
    fields: null,
    required_fields: ['installer_number'],
    optional_fields: ['environment', 'change_window', 'installation_notes'],
    display_name: 'Instalación en Pruebas',
    description: 'Instalación y configuración en ambiente de pruebas'
  }
};

// Definición de campos del formulario
export interface FormField {
  name: string;
  label: string;
  type: 'text' | 'number' | 'date' | 'select' | 'textarea';
  required: boolean;
  placeholder?: string;
  options?: { value: string; label: string }[];
  description?: string;
}

export interface StageFormConfig {
  stage_name: string;
  fields: FormField[];
}

File: backend/app/schemas/activity_log.py
"""
Esquemas para el log de actividades de desarrollo
"""

from pydantic import BaseModel, Field
from typing import Optional, Dict, Any, List
from datetime import date, datetime
from enum import Enum


class ActivityType(str, Enum):
    NUEVA_ACTIVIDAD = "nueva_actividad"
    SEGUIMIENTO = "seguimiento"
    CIERRE_ETAPA = "cierre_etapa"


class ActivityStatus(str, Enum):
    PENDIENTE = "pendiente"
    EN_CURSO = "en_curso"
    COMPLETADA = "completada"
    CANCELADA = "cancelada"


class ActorType(str, Enum):
    EQUIPO_INTERNO = "equipo_interno"
    PROVEEDOR = "proveedor"
    USUARIO = "usuario"


# Campos dinámicos por etapa
class AprobacionPropuestaFields(BaseModel):
    """Campos específicos para etapa de Aprobación de Propuesta"""
    proposal_id: str = Field(..., description="Número/ID de la propuesta")
    proposal_version: Optional[str] = Field(None, description="Versión de la propuesta")
    approver: Optional[str] = Field(None, description="Persona que aprueba")
    approval_date: Optional[date] = Field(None, description="Fecha de aprobación")
    budget_approved: Optional[float] = Field(None, description="Presupuesto aprobado")


class AnalisisFields(BaseModel):
    """Campos específicos para etapa de Análisis"""
    side_case_id: Optional[str] = Field(None, description="Número de caso SIDE")
    servicepoint_case_id: Optional[str] = Field(None, description="Número de caso ServicePoint")
    analyst_assigned: Optional[str] = Field(None, description="Analista asignado")
    scope_analysis: Optional[str] = Field(None, description="Alcance del análisis")
    requirements_count: Optional[int] = Field(None, description="Número de requerimientos")


class InstalacionPruebasFields(BaseModel):
    """Campos específicos para etapa de Instalación en Pruebas"""
    installer_number: str = Field(..., description="Número del instalador")
    environment: Optional[str] = Field(None, description="Ambiente de pruebas")
    change_window: Optional[str] = Field(None, description="Ventana de cambio")
    installation_notes: Optional[str] = Field(None, description="Notas de instalación")
    version: Optional[str] = Field(None, description="Versión del instalador")
    failure_type: Optional[str] = Field(None, description="Tipo de falla si aplica")
    resolution_attempts: Optional[str] = Field(None, description="Intentos de resolución")
    next_actions: Optional[str] = Field(None, description="Próximas acciones")


class ValidacionCorreccionesFields(BaseModel):
    """Campos específicos para etapa de Validación de Correcciones"""
    installer_number: str = Field(..., description="Número del instalador que se está validando")
    failure_description: str = Field(..., description="Descripción detallada de la falla encontrada")
    original_stage_reference: Optional[str] = Field(None, description="Etapa original donde se detectó la falla")
    correction_requirements: Optional[str] = Field(None, description="Requerimientos específicos de corrección")
    validation_notes: Optional[str] = Field(None, description="Notas de la validación realizada")
    provider_response: Optional[str] = Field(None, description="Respuesta del proveedor a las correcciones")
    correction_status: Optional[str] = Field(None, description="Estado de la corrección (pendiente, en_progreso, completada)")
    expected_correction_date: Optional[date] = Field(None, description="Fecha esperada de corrección")
    validation_result: Optional[str] = Field(None, description="Resultado de la validación (aprobada, rechazada, pendiente)")


class DevelopmentActivityLogBase(BaseModel):
    """Esquema base para actividades de desarrollo"""
    stage_id: int = Field(..., description="ID de la etapa")
    activity_type: ActivityType = Field(..., description="Tipo de actividad")
    start_date: date = Field(..., description="Fecha de inicio")
    end_date: Optional[date] = Field(None, description="Fecha de fin")
    next_follow_up_at: Optional[date] = Field(None, description="Próximo seguimiento")
    status: ActivityStatus = Field(ActivityStatus.PENDIENTE, description="Estado de la actividad")
    actor_type: ActorType = Field(..., description="Tipo de actor responsable")
    notes: Optional[str] = Field(None, description="Notas adicionales")
    dynamic_payload: Optional[Dict[str, Any]] = Field(None, description="Campos específicos por etapa")


class DevelopmentActivityLogCreate(DevelopmentActivityLogBase):
    """Esquema para crear una nueva actividad"""
    pass


class DevelopmentActivityLogUpdate(BaseModel):
    """Esquema para actualizar una actividad"""
    end_date: Optional[date] = None
    next_follow_up_at: Optional[date] = None
    status: Optional[ActivityStatus] = None
    notes: Optional[str] = None
    dynamic_payload: Optional[Dict[str, Any]] = None


class DevelopmentActivityLogResponse(DevelopmentActivityLogBase):
    """Esquema de respuesta para actividades"""
    id: int
    development_id: str
    created_by: Optional[str]
    created_at: datetime
    updated_at: Optional[datetime]
    
    # Información de la etapa
    stage_name: Optional[str] = None
    stage_code: Optional[str] = None
    
    class Config:
        from_attributes = True


class ActivityLogListResponse(BaseModel):
    """Respuesta para listado de actividades"""
    activities: List[DevelopmentActivityLogResponse]
    total: int
    page: int
    size: int


# Configuración de campos por etapa
STAGE_FIELD_CONFIGS = {
    "Aprobación Propuesta": {
        "fields": AprobacionPropuestaFields,
        "required_fields": ["proposal_id"],
        "optional_fields": ["proposal_version", "approver", "approval_date", "budget_approved"]
    },
    "Análisis": {
        "fields": AnalisisFields,
        "required_fields": [],
        "optional_fields": ["side_case_id", "servicepoint_case_id", "analyst_assigned", "scope_analysis", "requirements_count"]
    },
    "Despliegue (Pruebas)": {
        "fields": InstalacionPruebasFields,
        "required_fields": ["installer_number"],
        "optional_fields": ["environment", "change_window", "installation_notes", "version", "failure_type", "resolution_attempts", "next_actions"]
    },
    "Validación de Correcciones": {
        "fields": ValidacionCorreccionesFields,
        "required_fields": ["installer_number", "failure_description"],
        "optional_fields": ["original_stage_reference", "correction_requirements", "validation_notes", "provider_response", "correction_status", "expected_correction_date", "validation_result"]
    }
}


def get_stage_field_config(stage_name: str) -> Optional[Dict[str, Any]]:
    """Obtener configuración de campos para una etapa específica"""
    return STAGE_FIELD_CONFIGS.get(stage_name)


def validate_dynamic_payload(stage_name: str, payload: Dict[str, Any]) -> Dict[str, Any]:
    """Validar payload dinámico según la etapa"""
    config = get_stage_field_config(stage_name)
    if not config:
        return payload
    
    # Validar campos requeridos
    for field in config["required_fields"]:
        if field not in payload or payload[field] is None:
            raise ValueError(f"Campo requerido '{field}' no encontrado para etapa '{stage_name}'")
    
    # Filtrar solo campos válidos
    valid_fields = config["required_fields"] + config["optional_fields"]
    filtered_payload = {k: v for k, v in payload.items() if k in valid_fields}
    
    return filtered_payload
